README.md
IF Name Generator
=================

The name generator creates randomized names based on a list of existing names.
The names are generated by taking the beginning part of one name and appending
the end part of another name.

The default list is comprised of interactive fiction titles, but the algorithm
works just as well for any titles.

The original generator and algorithm was created by David Fisher, who was 
inspired by the [Video Game Name Generator](http://videogamena.me) and extracted 
the name list from Baf's Guide (see the [rec.arts.int-fiction thread](http://groups.google.fi/group/rec.arts.int-fiction/browse_thread/thread/4428f0f75e24abe1/)).
The current list is from the [IFDB database dump](http://ifarchive.org/indexes/if-archiveXinfoXifdb.html)
(January 2017).

The name generator is online at http://nitku.net/if/namegenerator.


API
===

The list of names can also be requested as JSON or plain text.
 
Send a GET request to `http://nitku.net/if/namegenerator/` with any of the
following parameters:

**`format`**: Either `html`, `txt`, `json` or `jsonp`. If the format is `jsonp`, 
a `callback` parameter must be supplied, otherwise a JSON response is returned.
Default is `html`.

**`n`**: The amount of names to return. Must be between 1 and 500. Default is 40.

**`ifcomp`**: If supplied, every title has at least one part taken from the
given [IFComp](http://ifcomp.org) year's titles. Must be between 1995 and 2016. 

Example: 

```
http://nitku.net/if/namegenerator/?format=json&n=50&ifcomp=2015
```

The server has `Access-Control-Allow-Origin: *` header set so cross-origin
requests are possible.

The API doesn't have a hard request limit, but if you plan on using it in a way
that puts considerable stress on the server, please discuss it beforehand or
consider installing the script to your own server. Excess usage can cause an IP block.

CHANGES
Version 4.1 (2017-01-28):
 - Updated the game list
 - Fixed some titles with $ in them
 - Fixed PHP warnings about undefined variables

Version 4 (2015-11-06):
 - Refactoring
 - Updated the game list
 - Added API and JSON, JSONP and text formats
 - Invalid parameters sends a 400 error code

Version 3 (2009-10-17):
 - Added all the previous IFComps and mashing only between comp names

Version 2:
 - Made it so that the first part can't be "the", "a" or "an" alone
 - Added a debug command to see where the title is split (append ?debug at the end of the url to see it)
 - Minor adjustments to the site layout

output/html.php
<?php
header( 'Content-Type: text/html; charset=utf8' );
?><!DOCTYPE html>
<html lang="en">
<head>
	<title>IF Name Generator</title>
	<meta charset="UTF-8">
	<meta description="IF Name Generator mashes two existing Interactive Fiction titles together to create a new one.">
	<meta http-equiv="Pragma" content="no-cache">
	<meta name="viewport" content="width=device-width, initial-scale=1">

	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css" integrity="sha512-dTfge/zgoMYpP7QbHy4gWMEGsbsdZeCXz7irItjcC3sPUFtf0kuFbDz/ixG7ArTxmDjLXDmezHubeNikyKGVyQ==" crossorigin="anonymous">
	<link rel="stylesheet" href="style.css" type="text/css">
</head>
<body>

<?php
if( $error ) {
    die( $error.'</body></html>' );
}
?>

<div class="container">
<p>
Below is a<?php if( $n > 1 ) {
    echo " list of " . $n;
} ?> randomized IF
title<?php if( $n > 1 ) { echo 's'; } ?> created by
joining the beginning and end parts of random existing titles.
</p>

<p>
Click on the title<?php if( $n > 1 ) { echo 's'; } ?> to read randomly generated reviews.
Reload the page to get a new <?php if( $n > 1 ) { echo 'set of titles'; } else { echo 'title'; } ?>.
</p>

<form action="" method="get" id="ifcomp-form" class="form-inline">
<p class="form-group">
Try the <a href="http://www.ifcomp.org/">IFComp</a> editions:
<select name="ifcomp" id="ifcomp-selection" class="form-control">
<option></option>
<?php
for( $loop = 1995; $loop <= end( (array_keys( $comps )) ); ++$loop ) {
?>
    <option value="<?php echo $loop; ?>"<?php if( $compEdition == $loop ) { echo ' selected'; } ?>><?php echo $loop; ?></option>
<?php
}
?>
<option value="all"<?php if( $compEdition == 'all' ) { echo ' selected'; } ?>>All IFComps</option>
</select>
<button type="submit" class="btn btn-primary" id="ifcomp-go">Go</button>
</p>
</form>
<script>
    document.getElementById( 'ifcomp-go' ).style.display = 'none';
    document.getElementById( 'ifcomp-selection' ).addEventListener( 'change', function() {
        document.getElementById( 'ifcomp-form' ).submit();
    }, false );
</script>
<hr />

<ul id="list">
<?php
for( $i = 0; $i < count( $generatedNames ); ++$i ) {
    echo '<li><a href="http://nitku.net/if/reviewgenerator/index.php?name='.urlencode( $generatedNames[ $i ] ).'" rel="nofollow">';
    echo $generatedNames[ $i ];
	echo "</a></li>\n";
}
?>
</ul>

</div>

<footer class="text-right small container">
<a href="https://github.com/juhana/name-generator">Source & API</a>
</footer>
</body>
</html>

output/jsonp.php
<?php
if( !empty( $_GET[ 'callback' ] ) ){
    $callback = $_GET[ 'callback' ];

    if( preg_match( '/[^0-9a-zA-Z\$_]|^(abstract|boolean|break|byte|case|catch|char|class|const|continue|debugger|default|delete|do|double|else|enum|export|extends|false|final|finally|float|for|function|goto|if|implements|import|in|instanceof|int|interface|long|native|new|null|package|private|protected|public|return|short|static|super|switch|synchronized|this|throw|throws|transient|true|try|typeof|var|volatile|void|while|with|NaN|Infinity|undefined)$/', $callback ) ) {
        die( 'Invalid callback identifier. Only alphanumeric characters, _ and $ allowed.' );
    }

    header( 'Content-Type: text/javascript; charset=utf8' );

    echo $callback . '(' . json_encode( $generatedNames ) . ');';
}
else {
    require "json.php";
}

index.php
<?php
// remove these two lines to disable cross-origin requests
header( 'Access-Control-Allow-Origin: *' );
header( 'Access-Control-Allow-Methods: GET, POST' );

define( 'DEFAULT_NAMES_TO_PRINT', 40 );
define( 'MAX_NAMES_TO_PRINT', 500 );

$OUTPUT_FORMATS = array( 'html', 'json', 'txt' );

if( isset( $_GET[ 'n' ] ) && ( !empty( $_GET[ 'n' ] ) || $_GET[ 'n' ] === '0' ) ) {
	$n = intval( $_GET[ 'n' ] );
}
else {
	$n = DEFAULT_NAMES_TO_PRINT;
}

$error = false;

if( $n < 1 || $n > MAX_NAMES_TO_PRINT ) {
    $error = "Number of titles must be between 1 and " . MAX_NAMES_TO_PRINT . '.';
}

if( empty( $_GET[ 'format' ] ) ) {
    $format = 'html';
}
else {
    $format = strtolower( trim( $_GET[ 'format' ] ) );
}

if( array_search( $format, $OUTPUT_FORMATS ) === false ) {
    $error = "Unknown output format '" . htmlspecialchars( $format ) . "'. Available formats: " . implode( $OUTPUT_FORMATS, ', ' );
}

require_once( 'names.php' );

if( !$error ) {
    if( $format === 'html' ) {
        debug( '<pre>' );
    }

    $compEdition = $_GET[ 'ifcomp' ];

    if( isset( $_GET[ 'edition' ] ) )
    {
        switch( $_GET[ 'edition' ] )
        {
            // for backwards compatibility with old links
            case 'comp08':
                $compEdition = 2008;
            break;

            case 'comp09':
                $compEdition = 2009;
            break;
        }
    }

    if( isset( $comps[ $compEdition ] ) )
    {
        $specialArray = $comps[ $compEdition ];
    }

    if( $compEdition == 'all' )
    {
        $names = array();

        foreach( $comps as $compArray )
        {
            $newNames = array_merge( $names, $compArray );
            $names = $newNames;
        }
    }

    $generatedNames = array();

    for( $loop = 0; $loop < $n; ++$loop )
    {
        do
        {
            $source = array_rand( $names, 2 );

            if( empty( $specialArray ) )
            {
                // no "special edition" selected, just pick two names
                $prefix = explode( ' ', $names[ $source[ 0 ] ] );
                $suffix = explode( ' ', $names[ $source[ 1 ] ] );
            }
            else
            {
                // othewise we'll take randomly either the end part
                // or the beginning part of the title from a special list
                if( rand( 0, 1 ) )
                {
                    $prefix = explode( ' ', $specialArray[ rand( 0, count( $specialArray ) - 1 ) ] );
                    $suffix = explode( ' ', $names[ $source[ 1 ] ] );
                }
                else
                {

                    $prefix = explode( ' ', $names[ $source[ 0 ] ] );
                    $suffix = explode( ' ', $specialArray[ rand( 0, count( $specialArray ) - 1 ) ] );
                }
            }


            $prefix_words = count( $prefix );
            $suffix_words = count( $suffix );

            // if the prefix is only one word, choose that. Otherwise
            // pick a random number of words from it.
            if( $prefix_words == 1 )
            {
                $prefixToPrint = $prefix[ 0 ] . ' ';
            }
            else
            {
                $prefix_pick = rand( 1, count( $prefix ) - 1 );
                $prefixToPrint = '';

                for( $i = 0; $i < $prefix_pick; ++$i )
                {
                    $prefixToPrint .= $prefix[ $i ] . ' ';
                }
            }

        // if the prefix is only "the" or "a" or something else
        // we don't want, loop through the algorithm until
        // we get a good result.
        } while( !words_ok( trim( $prefixToPrint ) ) );


        $fullname = $prefixToPrint;
        debug( $prefixToPrint . '| ' );

        if( $suffix_words == 1 )
        {
            $fullname .= $suffix[ 0 ];
            debug( $suffix[ 0 ] );
        }
        else
        {
            $suffix_pick = rand( 1, count( $suffix ) - 1 );

            for( $i = $suffix_pick; $i < $suffix_words; ++$i )
            {
                $fullname .= $suffix[ $i ] . ' ';
                debug( $suffix[ $i ] . ' ' );
            }
        }

        debug( "\n" );

        $generatedNames[ $loop ] = trim( $fullname );
    }

    if( $format === 'html' ) {
        debug( '</pre>' );
    }
}

function words_ok( $words )
{
    $badWords = array( 'the', 'a', 'an' );

    if( in_array( strtolower( $words ), $badWords ) )
    {
        return false;
    }

    return true;
}

function debug( $msg )
{
    if( isset( $_GET[ 'debug' ] ) )
    {
        echo $msg;
    }
}

if( $error ) {
    header( $_SERVER['SERVER_PROTOCOL'] . ' 400 Bad Request', true, 400 );
}

include "output/" . $format . ".php";